baseapp/fraudproof.go

```diff
@@ -0,0 +1,191 @@
+)
+
+// Represents a single-round fraudProof
+type FraudProof struct {
+	// The block height to load state of
+	blockHeight int64
+
+	// TODO: Add Proof that appHash is inside merklized ISRs in block header at block height
+
+	appHash []byte
+	// A map from module name to state witness
+	stateWitness map[string]StateWitness
+
+	// Fraudulent state transition has to be one of these
+	// Only one have of these three can be non-nil
+	fraudulentBeginBlock *abci.RequestBeginBlock
+	fraudulentDeliverTx  *abci.RequestDeliverTx
+	fraudulentEndBlock   *abci.RequestEndBlock
+
+	// TODO: Add Proof that fraudulent state transition is inside merkelizied transactions in block header
+}
+
+// State witness with a list of all witness data
+type StateWitness struct {
+	// store level proof
+	Proof    tmcrypto.ProofOp
+	RootHash []byte
+	// List of witness data
+	WitnessData []WitnessData
+}
+
+// Witness data containing a key/value pair and a SMT proof for said key/value pair
+type WitnessData struct {
+	Key   []byte
+	Value []byte
+	Proof tmcrypto.ProofOp
+}
+
+func (fraudProof *FraudProof) getModules() []string {
+	keys := make([]string, 0, len(fraudProof.stateWitness))
+	for k := range fraudProof.stateWitness {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+func (fraudProof *FraudProof) getDeepIAVLTrees() (map[string]*iavltree.MutableTree, error) {
+	storeKeyToIAVLTree := make(map[string]*iavltree.MutableTree)
+	for storeKey, stateWitness := range fraudProof.stateWitness {
+		rootHash := stateWitness.RootHash
+		// TODO(manav): Replace with IAVL Deep Subtrees once implementation is done
+		substoreDeepSMT := smtlib.NewDeepSparseMerkleSubTree(smtlib.NewSimpleMap(), smtlib.NewSimpleMap(), sha256.New(), rootHash)
+		for _, witnessData := range stateWitness.WitnessData {
+			proofOp, key, val := witnessData.Proof, witnessData.Key, witnessData.Value
+			proof, err := types.CommitmentOpDecoder(proofOp)
+			if err != nil {
+				return nil, err
+			}
+			// TODO(manav): Replace with an IAVL proof instead of SMT here
+			smtProof := proof.(*smt.ProofOp).GetProof()
+			substoreDeepSMT.AddBranch(smtProof, key, val)
+		}
+		// TODO(manav): Replace with actual iavl stores
+		storeKeyToIAVLTree[storeKey] = &iavltree.MutableTree{}
+	}
+	return storeKeyToIAVLTree, nil
+}
+
+// Returns true only if only one of the three pointers is nil
+func (fraudProof *FraudProof) checkFraudulentStateTransition() bool {
+	if fraudProof.fraudulentBeginBlock != nil {
+		return fraudProof.fraudulentDeliverTx == nil && fraudProof.fraudulentEndBlock == nil
+	}
+	if fraudProof.fraudulentDeliverTx != nil {
+		return fraudProof.fraudulentEndBlock == nil
+	}
+	return fraudProof.fraudulentEndBlock != nil
+}
+
+func (fraudProof *FraudProof) verifyFraudProof() (bool, error) {
+	if !fraudProof.checkFraudulentStateTransition() {
+		return false, fmt.Errorf("fraudProof has more than one type of fradulent state transitions marked nil")
```

Comment:
`errors.New` should be used instead of `fmt.Errorf`. This error should be extracted as package level sentinel error.